File: Done/BinSearch/P34.cc
#include <iostream>
#include <vector>
using namespace std;

vector<int> searchRange(vector<int> &nums, int target)
{
    if (nums.empty())
        return vector<int>({-1, -1});

    int mid = 0, l = 0, u = nums.size() - 1;
    while (l <= u)
    {
        mid = (l + u) / 2;
        if (nums[mid] == target)
            break;
        if (nums[mid] > target)
            u = mid - 1;
        else
            l = mid + 1;
    }

    if (l > u)
        return vector<int>({-1, -1});

    vector<int> sol(2, -1);
    l = mid;
    u = mid;

    while (l >= 0 && nums[l] == target) l--;
    while (u < nums.size() && nums[u] == target) u++;

    sol[0] = l + 1;
    sol[1] = u - 1;
    return sol;
}

int main()
{
    std::vector<int> nums;
    nums.push_back(5);
    nums.push_back(7);
    nums.push_back(7);
    nums.push_back(8);
    nums.push_back(8);
    nums.push_back(10);

    vector<int> sol = searchRange(nums, 8);

    cout << "Range: [" << sol[0] << ", " << sol[1] << "]" << std::endl;

    return 0;
}

File: Done/BinSearch/P34.java
package Done.BinSearch;

import java.util.Arrays;


public class P34 {

    static int[] searchRange(int[] nums, int target) {
        int mid, l = 0, u = nums.length - 1;
        int[] sol = {-1, -1};
        while (l < u) {
            mid = (l + u) / 2;
            if (nums[mid] == target) {
                u = mid;
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                u = mid - 1;
            }
        }
        if (target == nums[l]) {
            sol[0] = l;
        }
        l = 0;
        u = nums.length - 1;
        while (l < u) {
            mid = (int) Math.ceil((l + u) / 2.0);
            if (nums[mid] == target) {
                l = mid;
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                u = mid - 1; 
            }
        }
        if (target == nums[u]) {
            sol[1] = u;
        }
        return sol;
    }

    public static void main(String[] args) {
        int[] nums = {};
        System.out.println(Arrays.toString(searchRange(nums, 2)));
    }
}
